pkg/engine/imageVerifyMetadata.go

```diff
@@ -0,0 +1,96 @@
+package engine
+
+import (
+	"encoding/json"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"strings"
+)
+
+const imageVerifyAnnotationKey = "kyverno.io/images"
+
+type ImageVerificationMetadata struct {
+	Data map[string] bool `json:"data"`
+}
+
+func (ivm *ImageVerificationMetadata) add(image string, verified bool) {
+	if ivm.Data == nil {
+		ivm.Data = make(map[string]bool, 0)
+	}
+
+	ivm.Data[image] = verified
+}
+
+func (ivm *ImageVerificationMetadata) isVerified(image string) bool {
+	if ivm.Data == nil {
+		return false
+	}
+
+	verified, ok := ivm.Data[image]
+	if !ok {
+		return false
+	}
+
+	return verified
+}
+
+func parseImageMetadata(jsonData string) (*ImageVerificationMetadata, error) {
+	var data map[string]bool
+	if err := json.Unmarshal([]byte(jsonData), &data); err != nil {
+		return nil, err
+	}
+
+	return &ImageVerificationMetadata{
+		Data: data,
+	}, nil
+}
+
+func (ivm *ImageVerificationMetadata) Patches(hasAnnotations bool, log logr.Logger) ([][]byte, error) {
+	var patches [][]byte
+	if !hasAnnotations {
+		var addAnnotationsPatch = make(map[string]interface{})
+		addAnnotationsPatch["op"] = "add"
+		addAnnotationsPatch["path"] = "/metadata/annotations"
+		addAnnotationsPatch["value"] = map[string]string{}
+		patchBytes, err := json.Marshal(addAnnotationsPatch)
+		if err != nil {
+			return nil, err
+		}
+
+		log.V(4).Info("adding annotation patch", "patch", string(patchBytes))
+		patches = append(patches, patchBytes)
+	}
+
+	data, err := json.Marshal(ivm.Data)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to marshal metadata value: %v", data)
+	}
+
+	var addKeyPatch = make(map[string]interface{})
+	addKeyPatch["op"] = "add"
+	addKeyPatch["path"] = makeAnnotationKeyForJSONPatch()
+	addKeyPatch["value"] = string(data)
+
+	patchBytes, err := json.Marshal(addKeyPatch)
+	if err != nil {
+		return nil, err
+	}
+
+	log.V(4).Info("adding image verification patch", "patch", string(patchBytes))
+	patches = append(patches, patchBytes)
+	return patches, nil
+}
+
+func (ivm *ImageVerificationMetadata) Merge(other *ImageVerificationMetadata) {
+	for k, v := range other.Data {
+		ivm.add(k, v)
+	}
```

Comment:
Yes, that is correct. However, I think that this would be the appropriate behavior as Kyverno rules are independent.  

Favoring a negative result across rules, could also lead to situations where an image previously did not pass verification, but after the right attestations are added, passes.

Reply:
But in the current case we are favoring the rule that is evaluated last right? Do we want to document that somewhere?

Reply:
Its discussed here:

https://kyverno.io/docs/applying-policies/#in-clusters

> Rules in a policy are applied in the order of definition....

Reply:
@samj1912 - any additional thoughts or feedback, to merge this?